cirq-core/cirq/ops/conditional_operation.py

```diff
@@ -0,0 +1,161 @@
+# Copyright 2021 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import (
+    AbstractSet,
+    Any,
+    cast,
+    Dict,
+    Optional,
+    Sequence,
+    TYPE_CHECKING,
+    Tuple,
+    Union,
+)
+
+from cirq import protocols, value
+from cirq.ops import raw_types
+
+if TYPE_CHECKING:
+    import cirq
+
+
+@value.value_equality
+class ConditionalOperation(raw_types.Operation):
+    """Augments existing operations to be conditionally executed."""
+
+    def __init__(
+        self,
+        sub_operation: 'cirq.Operation',
+        controls: Sequence[Union[str, 'cirq.MeasurementKey']],
+    ):
+        controls = tuple(value.MeasurementKey(k) if isinstance(k, str) else k for k in controls)
+        if isinstance(sub_operation, ConditionalOperation):
+            # Auto-flatten nested controlled operations.
+            sub_operation = cast(ConditionalOperation, sub_operation)
+            self._controls = controls + sub_operation.controls
+            self._sub_operation = sub_operation.sub_operation
+        else:
+            self._controls = controls
+            self._sub_operation = sub_operation
+
+    @property
+    def controls(self):
+        return self._controls
+
+    @property
+    def sub_operation(self):
```

Comment:
Does this even need to be public? It could be private, then the "number of wrappers" problem is less troublesome. We can just design around the best way to make `op.with_tags().with_conditions().with_tags().without_conditions()....` behave the way we want. Plausibly even the `ConditionalOperation` class itself could be private.